numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -328,28 +328,31 @@ gentype_nonzero_number(PyObject *m1)
 static PyObject *
 gentype_str(PyObject *self)
 {
-    PyObject *arr, *ret = NULL;
+    PyObject  *item, *item_str;
 
-    arr = PyArray_FromScalar(self, NULL);
-    if (arr != NULL) {
-        ret = PyObject_Str((PyObject *)arr);
-        Py_DECREF(arr);
+    item = gentype_generic_method(self, NULL, NULL, "item");
+    if (item == NULL) {
+        return NULL;
     }
-    return ret;
+
+    item_str = PyObject_Str(item);
+    Py_DECREF(item);
+    return item_str;
 }
```

Comment:
Yes, the docs for user-defined-types say `In addition, the required functions in the ”.f” member must be defined: nonzero, copyswap, copyswapn, setitem, getitem, and cast.`.

They don't say if `self` is a valid return value, but I think it isn't since none of the numpy types do that.